Remove GLib dependency within bin/

mack-w opened the pull request:
See #133.

mack-w committed:
chore: move traceAnalyzer/threadPool.cpp to utils/

This commit is the first in a series of commits to remove gLib.

This commit moves an unused thread pool library within traceAnalyzer
to the utils directory, which may find its use elsewhere.

mack-w committed:
Revert "chore: move traceAnalyzer/threadPool.cpp to utils/"

This reverts commit 1155efd98708e0dfec479c88cb0457dbdaa6f43f.

mack-w committed:
feat(include): introduce sheredom/hashmap.h

This commit is the first in series (ignoring reverted) to remove gLib.

This commit introduces [hashmap.h](https://github.com/sheredom/hashmap.h)
in an effort to remove GHashTable dependencies from the project.

Future commits shall introduce similar header-only libs as well.

mack-w committed:
feat(bin/cachesim): drop GHashTable for header-only hash table lib

mack-w committed:
feat(bin/MRC): drop GHashTable for header-only hash table lib

mack-w committed:
fix(bin/MRC): struct field naming in commit f9d1c49

mack-w committed:
chore(include): move redundant defs of obj_id_hasher together

mack-w committed:
feat(bin/MRC): complete migration to header-only hash table lib

1a1a11a commented:
just FYI, it might be better to start with a small PR :)

mack-w commented:
Thanks for the suggestion. So let's begin with something like "remove GLib dependency from main executable" first? Since gLib usage within the algo part is more extensive and complicated, maybe I shall get rid of GLib in `bin/` first (and I'm close to that.)

BTW the quality gate check is complaining about duplicated code on `hashmap.h`... it's saying like "for i=0" is a duplicate. I guess I'm gonna ignore that unless there're other problems with cq.

mack-w committed:
feat(profiler): remove reference of g_mutex_*

mack-w committed:
WIP: feat(profiler): use homebrewed threadpool

Move from GThreadPool to a homebrewed threadpool
adapted from https://nachtimwald.com/2019/04/12/thread-pool-in-c

Note that this code **only** runs with UBSan enabled.
Not for production use (yet).

1a1a11a commented:
Sounds good!

mack-w committed:
fix: clang-format problem in GitHub Actions

mack-w committed:
feat(profiler): use homebrewed threadpool

Finalize move from GThreadPool to a homebrewed threadpool
adapted from https://nachtimwald.com/2019/04/12/thread-pool-in-c

Note, that simulation results differ from original.
Reason for this is under investigation.

mack-w commented:
@1a1a11a Hi! I would say that this PR is near finished.

In the meantime, I'm observing an at-most 1% difference in simulation results for **some** algorithms (e.g. FIFO) with `mrcProfiler`. I also see assertion error `hit_size_vec[9]` in `testMrcProfiler::test_minisim_profiler()`. I'm still investigating the reason for this. 

Right now sanitizers are not throwing any useful error, and vvverbose logs look good as well. Since `testMrcProfiler` is single-threaded, there shall be no data races. I will try to git bisect what's going wrong here. In the meantime if you have any hints or you want to see the logs please reply.

PS: that `strncpy` security hotspot emerged out of the blue and I `ASSERT_IT_IS_SAFE`

1a1a11a commented:
loop in @JasonGuo98

mack-w commented:
> loop in @JasonGuo98

Did a quick test, it seems that the thread pool library that I'm quoting is problematic. I switched to a battle-tested thread pool library, and (from preliminary test results) the problem is gone :)

Gotta do some git rebase afterwards...

~~For the deeper side of the problem, I suspect it's still the synchronisation problem with the reader that I'm not coping well with., since TSan reports a data race issue there. It is a benign problem at first glance, but who knows.~~ Pushing a new commit to fix this soon.

Upd: seems more like a problem with simulation steps, rather than reader problem. Introduced with the latest commit. Overlooked the problem since TSan keeps reporting the other way.

mack-w commented:
Add these two lines to respective functions in `simulator.c`:
```c
INFO("actual progress %d\n", progress);
INFO("actual num_of_caches %d\n", num_of_caches);
```

Before commit af3ff57:
```
[INFO]  simulator.c:307  (tid=126433726604224): actual progress 10
[INFO]  simulator.c:308  (tid=126433726604224): actual num_of_caches 10
ok 3 /libCacheSim/test_minisim_profiler_with_fixed_sample_rate
```

After commit af3ff57:
```
[INFO]  simulator.c:350  (tid=130706295354304): actual progress 9
[INFO]  simulator.c:351  (tid=130706295354304): actual num_of_caches 10
ERROR:test/test_mrcProfiler.cpp:188:void test_minisim_profiler_with_fixed_sample_rate(gconstpointer): assertion failed
not ok 3 /libCacheSim/test_minisim_profiler_with_fixed_sample_rate
```

The logic of the original code says:
```c
while (progress < num_of_caches - 1) {
    print_progress((double)progress / (double)(num_of_caches - 1) * 100);
}
```

~~It seems that *progress* shall be *num_of_caches* **minus** 1, but with the original code (upstream HEAD) it is not the case. Is it designed like this on purpose? Or is it just a data race problem, as TSan reported?~~

Update: it is designed for purpose, my bad. Pushing a fix.

Anyways, marking this PR as ready for review.

mack-w committed:
fix(simulator): incorrect simulation steps

The previous commit af3ff57 introduced pthread_cond_t to synchronize
progress. However, in some circumstances, the simulate_* functions may
exit prematurely due to incorrect conditional wait. This commit fix
the abovementioned problem.

1a1a11a commented on libCacheSim/bin/cli_reader_utils.c:
Can you add some documentation on how to use the data structures in the repo?

mack-w replied:
Actually `hashmap.h` is well documented. I shall then add code documentation and brief README for threadpool.h

1a1a11a replied:
It would be good to add the documentation in `docs` because we would like other contributors to use it as well. If it is hidden in the code, it might not be seen by others.

1a1a11a commented on libCacheSim/bin/cli_reader_utils.c:
```diff
@@ -267,11 +274,12 @@ void cal_working_set_size(reader_t *reader, int64_t *wss_obj,
     }
 
-    if (g_hash_table_contains(obj_table, (gconstpointer)req->obj_id)) {
+    if (hashmap_get(&new_obj_table, (const void *)(req->obj_id),
```
Can we just use obj_id as the key?

mack-w replied:
Yes, it is done as so. There is a custom hasher for `obj_id` (that I copied from gLib implementation.) Instead of treating everything as pointers as gLib does, this custom hasher takes the value of `obj_id`. Passing `obj_id` as pointer is just a requirement by function signature (since there may be a need to use other stuff than uint64 as key)

1a1a11a replied:
Similar to previous issue, let's use a custom macro to wrap the conversion so that it would be easier to change if need to migrate and refactor later.

mack-w commented:
Yeah, for sure. I will add some code comments. Do you want to have some separate documentation as well?

Best regards,
Mack Wang Y.
________________________________
From: Juncheng Yang ***@***.***>
Sent: Thursday, May 22, 2025 8:50:43 AM
To: 1a1a11a/libCacheSim ***@***.***>
Cc: Mack Wang ***@***.***>; Author ***@***.***>
Subject: Re: [1a1a11a/libCacheSim] Remove GLib dependency within bin/ (PR #184)


@1a1a11a commented on this pull request.

________________________________

On libCacheSim/bin/cli_reader_utils.c<https://github.com/1a1a11a/libCacheSim/pull/184#discussion_r2101418064>:

Can you add some documentation on how to use the data structures in the repo?

—
Reply to this email directly, view it on GitHub<https://github.com/1a1a11a/libCacheSim/pull/184#pullrequestreview-2859474586>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ARHUJDV2CVBKZVTDTBX2QCL27UNOHAVCNFSM6AAAAAB5GYESTOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDQNJZGQ3TINJYGY>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

1a1a11a commented on libCacheSim/profiler/threadpool.c:
```diff
@@ -0,0 +1,180 @@
+
+#include "../include/libCacheSim/threadpool.h"
+
+#include <pthread.h>
```
We should move threadpool to utils or dataStructure and add a quick README

mack-w replied:
Indeed

1a1a11a commented:
Yes, you can add a separate section for data structures in [doc](https://github.com/1a1a11a/libCacheSim/tree/develop/doc)

mack-w commented on libCacheSim/profiler/simulator.c:
```diff
@@ -29,29 +30,34 @@ typedef struct simulator_multithreading_params {
   sim_mt_params_t *params = (sim_mt_params_t *)user_data;
-  int idx = GPOINTER_TO_UINT(data) - 1;
+  // int idx = GPOINTER_TO_UINT(data) - 1;
+  int idx = ((unsigned int)(unsigned long)(data)) - 1;
```
Either we pass pointer to `int`, or we use `size_t` in place of `unsigned long`.
If we don't care about portability to 32-bit platforms or non amd64/aarch64, we just ignore this.
@1a1a11a What do you think?

1a1a11a replied:
Can we use an inline function or macro for this? We can ignore the portability for now, but if we need to revisit this later, changing the macro would be easier than changing it each time this conversion occurs.

mack-w commented:
Just noted that there seems to be a deadlock problem that I didn't notice while pushing. Trying to fix

mack-w committed:
fix(simulator): deadlock in simulate_* functions

In Release builds, access to params->n_caches is optimized to a stack
access. In Debug builds however, access is by pointer. Failure to
initialize this leads to infinite loop in _simulate, thus deadlock in
simulate_* functions.

mack-w committed:
chore: move threadpool to utils/

mack-w committed:
doc: some code comments for threadpool.h and hashmap.h

mack-w committed:
doc: documentation for thread pool and hash table

1a1a11a commented on doc/datastructures.md:
```diff
@@ -0,0 +1,275 @@
+- Use: Generic (non-cache related)
+- Path: *include/libCacheSim/hashmap.h*
+
+This is a generic hash table imported from [sheredom/hashmap.h](https://github.com/sheredom/hashmap.h). It is used everywhere except with cache objects, where there exists an optimised hashtable implementation.
```
"optimized"

1a1a11a commented on libCacheSim/bin/MRC/SHARDS.c:
```diff
@@ -20,44 +20,54 @@ extern "C" {
+         params->prio_tree->value >= params->threshold) {
     struct key *max = find_max_t(params->prio_tree)->key;
     uint64_t last_max = 0;
     params->rate = (double)max->Tmax / (double)(1 << 24);
```
debugging information?

mack-w replied:
🤔 didn't quite get what you mean, do I need to add some `DEBUG()` here?

1a1a11a replied:
Oh, when I read it yesterday, it was about some code that was commented out. Just want to remind you to remove unneeded code before merge. If some debugging-related code is needed and you want to keep the code, it is also ok as long as you add a comment stating why it is kept.

1a1a11a commented on libCacheSim/bin/MRC/SHARDS.c:
```diff
@@ -20,44 +20,54 @@ extern "C" {
-      g_hash_table_remove(params->lookup_hash, GSIZE_TO_POINTER((gsize)id));
-      params->distance_tree = splay_delete((long long)hash_value_inner, params->distance_tree);
+      void *hash_value_inner =
+          hashmap_get(params->lookup_hash, (const void *)id, sizeof(obj_id_t));
```
Replace the conversion with macro?

mack-w replied:
Yeah, I shall define a new header file to accommodate these common GLib macros

1a1a11a commented on libCacheSim/include/libCacheSim/hashmap.h:
[optional] it would be good to have a benchmark for hash table.

mack-w replied:
Shall figure out a way to do so
Does it need to be a benchmark on libCacheSim, or a independent/use case-irrelevant benchmark?

1a1a11a replied:
An independent benchmark should be sufficient, and it would be good to compare with a common library, e.g., Glib, C++ std, or absl hashtable. BTW, I am also creating a new issue to build a performance test for libCacheSim.

mack-w replied:
work in progress
@1a1a11a Where shall I put the benckmark? As a test program or just detail in the docs

1a1a11a commented on libCacheSim/profiler/dist.c:
```diff
@@ -6,13 +6,16 @@
 #include "../dataStructure/splay.h"
-#include "../include/libCacheSim/dist.h"
+#include "../include/libCacheSim/hashmap.h"
+#include "../include/libCacheSim/hashmap_defs.in"
```
We should use `.h` for header files, `.in` is for pre-processing

1a1a11a commented on libCacheSim/profiler/simulator.c:
```diff
@@ -184,23 +198,35 @@ cache_stat_t *simulate_at_multi_sizes(reader_t *reader, const cache_t *cache, in
-  GThreadPool *gthread_pool = g_thread_pool_new((GFunc)_simulate, (gpointer)params, num_of_threads, TRUE, NULL);
-  ASSERT_NOT_NULL(gthread_pool, "cannot create thread pool in simulator\n");
+
+  // GThreadPool *gthread_pool = g_thread_pool_new((GFunc)_simulate,
```
clean up

mack-w replied:
my bad, will clean these up together with resolving code formatting issues

mack-w commented:
@1a1a11a Sorry for not working this week, having some personal issues that I have to fix. Shall resume next week

mack-w committed:
fix: clang-format problem (again) in GitHub Actions

mack-w committed:
fix: various issues raised within PR comments

Includes comments cleanup, formatting issues, FIXME notes on conversion
macros (in next commit), spelling, etc.

mack-w committed:
fix: didn't include hashmap_defs.h in prev commit ef86c83

mack-w committed:
chore: introduce macro for common conversions e.g. int to pointer

mack-w committed:
fix: didn't include cli_reader_utils in prev commit

1a1a11a commented:
You can create a separate benchmark folder in the top directory, i.e., `benchmarks/dataStructure/`, would this work?

mack-w commented:
That makes sense considering that you want to add benchmark to the simulator.
Anyways I'm putting this as second in priority list, maybe I will open a separate PR for this :)

1a1a11a commented:
Shall I take a look?

mack-w commented:
> Shall I take a look?

Yeah I consider this PR "finished" (except for the benchmark part)
@1a1a11a Sorry I really had a lot of personal issues recently... I think next week I'll finally be able to look into smth with focus.

1a1a11a commented:
@haochengxia Can you take a look at this PR?

1a1a11a committed:
Merge branch 'develop' into develop

haochengxia commented on libCacheSim/profiler/dist.c:
```diff
@@ -85,7 +91,7 @@ int64_t get_stack_dist_add_req(const request_t *req, sTree **splay_tree,
     newtree = insert(curr_ts, *splay_tree);
   } else {
     // not first time access
```
When converting a pointer to an integer, can we use `int64_t old_ts = (int64_t)(uintptr_t)gp;` to avoid potential issues across archs?

BTW, as we removed glib, is `gp` still a good variable name?

haochengxia commented on libCacheSim/include/conversion.h:
```diff
@@ -0,0 +1,40 @@
+#define cptr_t_ const void*
+
+// platform-depedent
+#define ptr_size_t_ size_t
```
It is now quite good, and I recommend using `stdint.h` for robustness and readability.

e.g.,
#define ptr_uintptr_t_ uintptr_t
#define ptr_intptr_t_ intptr_t

#define int_to_ptr(i) (ptr_t_)((ptr_intptr_t_)i)
#define uint_to_ptr(i) (ptr_t_)((ptr_uintptr_t_)i)
#define int_to_cptr(i) (cptr_t_)((ptr_intptr_t_)i)
#define uint_to_cptr(i) (cptr_t_)((ptr_uintptr_t_)i)

#define ptr_to_int(p) (int)((ptr_intptr_t_)p)
#define ptr_to_uint(p) (unsigned int)((ptr_uintptr_t_)p)
...

haochengxia commented on libCacheSim/profiler/profilerLRU.c:
```diff
@@ -6,15 +6,18 @@
-#include "../dataStructure/splay.h"
 #include "../include/libCacheSim/profilerLRU.h"
 
+#include "../dataStructure/splay.h"
```
We can preserve the order of header files.

mack-w replied:
I'd like to preserve the order, but the current order is determined by `clang-format`.

mack-w commented:
@haochengxia Thanks for the code suggestions.
Also I see there's merge conflict. I will fix that and update my own branch soon.

mack-w committed:
Merge remote-tracking branch 'upstream/develop' into develop

mack-w commented:
Closed upon maintainers' decision.
